File: population.js
class Population {
    constructor(n, s, walls, checkpoints) {
        this.steps = s;
        this.n = n;
        this.walls = walls;
        this.checkpoints = checkpoints;
        this.cars = new Array(n);
        this.neat = new Neat(8, 4, n);
        this.matpool = [];
        for (let i = 0; i < this.n; ++i) {
            this.cars[i] = new Car(20, 40, this.walls, this.neat.genoms[i]);
        }
        this.alive = new Array(n);
        this.alive.fill(true);
    }

    checkColision(car) {
        for (let i = 0; i < this.walls.length; ++i) {
            if (car.collides(this.walls[i])) {
                return true;
            } 
        }
        return false;
    }

    progress(car, step) {
        for (let i = 0; i < this.checkpoints.length; ++i) {
            let frontWall = car.walls[0];
            if (frontWall.intersection(this.checkpoints[i])) {
                if (car.checkpoints.length == i) {
                    car.checkpoints.push([i, step]);
                }
            } 
        }
    }

    run(s) {
        let end = true;
        // For every car
        for (let i = 0; i < this.n; ++i) {
            if (this.alive[i]) {
                end = false;
                this.cars[i].think();
                this.cars[i].move();
                this.cars[i].show("green");
                this.cars[i].checkVel();
                this.progress(this.cars[i], s);
            }
            else this.cars[i].show("red");

            if (this.checkColision(this.cars[i])) {
                this.alive[i] = false;
                this.cars[i].deadAt = s;
            }
        }
        if (end) return Infinity;
        return ++s;
    }

    selection() {
        let fitnessArray = new Array(this.cars.length);
        let maxFitness = 0;
        let maxFitCar = 0;
        for (let i = 0; i < this.cars.length; ++i) {
            let aux = this.cars[i].fitness();
            fitnessArray[i] = aux;
            if (aux > maxFitness) {
                maxFitness = aux;
                maxFitCar = i;
            }
        }
        this.maxFitCar = maxFitCar;
        
        this.matpool = [];
        // Normalize weights and create matpool
        for (let i = 0; i < fitnessArray.length; ++i) {
            let vegades = fitnessArray[i]*1000/maxFitness;
            for (let x = 0; x < vegades; ++x) {
                this.matpool.push(i);
            }
        }
        console.log("Max fitness: " + maxFitness)
    }

    reproduction() {
        let bestBrain = this.cars[this.maxFitCar].brain;
        bestBrain.print();
        
        for (let i = 0; i < this.n; ++i) {
            let i1 = Math.floor(Math.random()*this.matpool.length);
            let i2 = Math.floor(Math.random()*this.matpool.length);
            
            let pare = this.cars[this.matpool[i1]];
            let mare = this.cars[this.matpool[i2]];
            this.cars[i].brain = pare.crossover(mare);
            
            this.cars[i].brain.mutate();
            this.cars[i].restart();
        }
        // Tots els cotxes tornen a estar vius 
        this.alive.fill(true);
    }
}

File: drift-v2/population.py
from car import Car
import math
from random import random
from neat import Neat

class Population():
    def __init__(self, n, s, walls, checkpoints):
        self.steps = s
        self.n = n
        self.walls = walls
        self.checkpoints = checkpoints
        self.cars = []
        self.neat = Neat(8, 4, n, self) # TODO: mirar això si ho faré així o com
        self.fitness_array = [0]*self.n
        self.matpool = []
        
        for i in range(n):
            self.cars.append(Car(20, 40, walls, self.neat.genoms[i]))
        
        self.alive = [True] * n

    def check_colision(self, car):
        for w in self.walls:
            if car.collides(w):
                return True
        
        return False

    def progress(self, car, step):
        for i, c in enumerate(self.checkpoints):
            front_wall = car.walls[0]
            check = False
            for w in car.walls:
                if w.intersection(c) and len(car.checkpoints) == i: check = True
            if check: car.checkpoints.append([i, step])

    
    def run(self, step):
        end = True
        for i, car in enumerate(self.cars):
            if self.alive[i]:
                end = False
                car.think()
                car.move()
                car.check_vel()
                self.progress(car, step)

            if self.check_colision(car):
                self.alive[i] = False
                car.dead_at = step
        
        if end: return math.inf
        step += 1
        return step

    def selection(self):
        max_fitness = -1
        max_fit_car = 0
        for i in range(self.n):
            aux = self.cars[i].fitness()
            self.fitness_array[i] = aux
            if aux > max_fitness:
                max_fitness = aux
                max_fit_car = i
        
        self.max_fit_car = max_fit_car
        self.matpool = []

        # # Normalize weights and create matpool
        # for i in range(len(self.fitness_array)):
        #     times = int(self.fitness_array[i] * 1000/max_fitness)
        #     for x in range(times):
        #         self.matpool.append(i)

    def reproduction(self):
        self.neat.reproduction(self.fitness_array)

        for i in range(self.n):
            self.cars[i].brain = self.neat.genoms[i]
            self.cars[i].restart()
        self.alive = [True] * self.n
        